crates/yttrium/src/sign/mod.rs

```diff
@@ -0,0 +1,1503 @@
+// - State:
+//   - is app and wallet state coupled? should we build the DApp support right now to make it easier?
+//   - does deduplication happen at the irn_subscription layer (like current SDKs) or do we do it for each action e.g. update, event, etc. (remember layered state and stateless architecture)
+
+// TODO
+// - WS reconnection & retries
+//   - disconnect if no ping for 30s etc. (native only)
+//   - reconnect back-off w/ random jitter to prevent server overload
+//   - web online/offline hints
+//   - callbacks for native hints?
+
+// TODO
+// - session pings, update, events, emit
+// - session expiry & renew
+//   - expire implemented simply by filtering out expired sessions in `Client::add_sessions()` ?
+//     - long-lived clients might suffer here. Maybe filter each reconnect?
+
+// TODO error improvement
+// - bundle size optimization: error enums only for actionable errors higher-up
+// - use a single string variant for all errors (which would be shown to users!)
+// - other is internal errors we don't expect to EVER happen (so show error code instead w/ GitHub issue link)
+// TODO bundle size optimization
+// - use native crypto utils
+// TODO relay changes
+// - subscribe to other sessions as part of `wc_approveSession` etc.
+// - (feasible?) wc_sessionRequestRespond which ACKs the `irn_subscription` message
+// - https://www.notion.so/walletconnect/Design-Doc-Sign-Client-Rust-port-2303a661771e80628bdbf07c96a97b21?source=copy_link#2303a661771e808f895acbcab46bd12a
+// - don't forward ignored messages e.g. ACKing etc. do it based on client version/flag
+// - binary relay encoding: bincode?
+
+// TODO
+// - Verify API
+// - 1CA
+// - Link Mode
+// - Events SDK & Analytics/TVF
+//   - Additional events for measuring latency/reconnect performance/client network environment/etc. so we can tune. E.g. "should we retry to connect?"
+// - Network state hinting (offline/online)
+//   - offline: don't try to reconnect, but also don't force a disconnect
+//   - online: reconnect if online() was called
+
+#[allow(unused)]
+impl Client {
+    pub fn new(
+        project_id: ProjectId,
+    ) -> (
+        Self,
+        tokio::sync::mpsc::UnboundedReceiver<(Topic, SessionRequestJsonRpc)>,
+    ) {
+        assert_eq!(
+            project_id.value().len(),
+            32,
+            "Project ID must be exactly 32 characters"
+        );
+
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+        (
+            Self {
+                relay_url: RELAY_URL.to_string(),
+                project_id,
+                key: None,
+                websocket: None,
+                session_request_tx: tx,
+                sessions: Arc::new(RwLock::new(HashMap::new())),
+                invalid_auth: Arc::new(AtomicBool::new(false)),
+            },
+            rx,
+        )
+    }
+
+    pub fn set_key(&mut self, key: SecretKey) {
+        self.key = Some(SigningKey::from_bytes(&key));
+    }
+
+    pub fn add_sessions(
```

Comment:
When should native SDK call it?

Reply:
Same with `key`, before you call `online()` or another function e.g. `pair()` or `approve()`. This initializes the state. You could initialize the state at the same time you construct the client, or later. Up to you